quay/src/lib.rs

```diff
@@ -8,4 +9,20 @@ extern crate serde;
 extern crate serde_derive;
 extern crate serde_json;
 
+use std::io::BufRead;
+use std::path::PathBuf;
+
 pub mod v1;
+
+pub fn read_credentials(credentials_path: Option<&PathBuf>) -> failure::Fallible<Option<String>> {
+    match &credentials_path {
+        Some(pathbuf) => match std::io::BufReader::new(std::fs::File::open(pathbuf)?)
```

Comment:
Nit: this is a quite terse line and the whole function is pretty heavy on `Option`s matching.
I would personally make it more straight and focused on the positive case by:
 * passing a pathbuf directly (non-optional)
 * splitting the file opening on its own, and adding context for the error
 * introduce some `use`s to reduce visual cluttering
 * reduce the amount of `match`es
 * taking care of a possibly empty token at the end

That is, something like:
```rust
let fp = fs::File::open(pathbuf).context(...)?;
let first_line = io::BufReader::new(fp).lines().nth(0)
  .ok_or_else(|| "empty credentials.")?;
let token = first_line?.trim_end().to_string();

if !token.is_empty() {
  Ok(Some(token))
} else {
  Ok(None)
}
```

Reply:
That's way too many lets for my taste :smirk: I think we've arrived in opinion-land

Reply:
To be clear, this wasn't just an opinion. This function is missing context on file-opening error and a final check on empty string, but I was worried you would go on nesting things here. Current code is artificially hard to review and follow, because the positive flow is implicit and mixed/split across match arms (while the underlying logic is linear).

It is fine to have separate `let` to split apart logical entities. You could collapse all `let`s in my example above if really wanted, but the goal should be to have it easy to read/maintain/debug, not to write as much FP as possible (which could be done by sprinkling a few more `map`/`and_then`, but I wouldn't recommend that).

Reply:
Thanks for the hint on the missing checks. I restructured the code based on your suggestion with more error context. I've also added an error for an empty token, since that's surely not going to work, and in case a file is provided it's probably not intended either.

Reply:
Ack, I agree that's more sensible.